// react-dadata/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpCache = exports.EmailSuggestions = exports.FioSuggestions = exports.BankSuggestions = exports.PartySuggestions = exports.AddressSuggestions = void 0;
var AddressSuggestions_1 = require("./AddressSuggestions");
Object.defineProperty(exports, "AddressSuggestions", { enumerable: true, get: function () { return AddressSuggestions_1.AddressSuggestions; } });
var PartySuggestions_1 = require("./PartySuggestions");
Object.defineProperty(exports, "PartySuggestions", { enumerable: true, get: function () { return PartySuggestions_1.PartySuggestions; } });
var BankSuggestions_1 = require("./BankSuggestions");
Object.defineProperty(exports, "BankSuggestions", { enumerable: true, get: function () { return BankSuggestions_1.BankSuggestions; } });
var FioSuggestions_1 = require("./FioSuggestions");
Object.defineProperty(exports, "FioSuggestions", { enumerable: true, get: function () { return FioSuggestions_1.FioSuggestions; } });
var EmailSuggestions_1 = require("./EmailSuggestions");
Object.defineProperty(exports, "EmailSuggestions", { enumerable: true, get: function () { return EmailSuggestions_1.EmailSuggestions; } });
var http_cache_1 = require("./http-cache");
Object.defineProperty(exports, "HttpCache", { enumerable: true, get: function () { return http_cache_1.HttpCache; } });
